### PR TITLE
Add UWS support for handling websocket frame fragmentation

### PR DESCRIPTION
**Pull request by blico**
Fix for issue #154 

uws_client was not handling frame fragmentation, resulting in the dropping of fragmented frames:

- SRS_UWS_CLIENT_01_213: [ A fragmented message consists of a single frame with the FIN bit clear and an opcode other than 0, followed by zero or more frames with the FIN bit clear and the opcode set to 0, and terminated by a single frame with the FIN bit set and an opcode of 0. ]

- SRS_UWS_CLIENT_01_220: [ Clients and servers MUST support receiving both fragmented and unfragmented messages. ]

Fixed by buffering fragmented frames, and not bubbling up to user until the frame is fully received. 
Tested by ensuring it fixes our broken scenario, and added unit tests to uws_client_ut.

### CODE REVIEW

File: src/uws_client.c [hunk 1]
```diff
@@ -1356,10 +1477,16 @@ static void on_underlying_io_bytes_received(void* context, const unsigned char*
                                 size_t pong_frame_length;
                                 BUFFER_HANDLE pong_frame_buffer;
 
-                                uws_client->uws_state = UWS_STATE_ERROR;
```
Comment (blico) [inline src/uws_client.c]: Worth calling out that I removed this line. It seems like a bug that we are setting the state to error when receiving a ping frame, but I could be missing something.

File: src/uws_client.c [hunk 2]
```diff
@@ -1184,54 +1185,173 @@ static void on_underlying_io_bytes_received(void* context, const unsigned char*
...
+                                    else
+                                    {
+                                        uws_client->fragment_buffer = new_fragment_bytes;
+                                        (void)memcpy(uws_client->fragment_buffer + uws_client->fragment_buffer_count, uws_client->stream_buffer + needed_bytes - length, length);
+                                        uws_client->fragment_buffer_count = length;
+                                    }
+                                    /* Codes_SRS_UWS_CLIENT_01_225: [ As a consequence of these rules, all fragments of a message are of the same type, as set by the first fragment's opcode. ]*/
+                                    /* Codes_SRS_UWS_CLIENT_01_226: [ Since control frames cannot be fragmented, the type for all fragments in a message MUST be either text, binary, or one of the reserved opcodes. ]*/
+                                    uws_client->fragmented_frame_type = WS_FRAME_TYPE_TEXT;
+                                }
```
Comment (jebrando) [inline src/uws_client.c]: This section of code is repeated below for binary, can you please create a function to reduce duplication.

File: src/uws_client.c (hunk 1 above)
Comment (jebrando) [reply src/uws_client.c]: I talked to the implementer and you are correct.

---
In reply to: [188471094](https://github.com/Azure/azure-c-shared-utility/pull/156#discussion_r188471094) [](ancestors = 188471094)

File: tests/uws_client_ut/uws_client_ut.c [hunk 3]
```diff
@@ -3932,6 +3937,319 @@ TEST_FUNCTION(when_a_0_bytes_text_frame_is_received_it_shall_be_indicated_to_the
...
+    // act
+    g_on_bytes_received(g_on_bytes_received_context, test_stream, sizeof(test_stream));
+    // assert
+    ASSERT_ARE_EQUAL(char_ptr, umock_c_get_expected_calls(), umock_c_get_actual_calls());
+
+    // cleanup
+    uws_client_destroy(uws_client);
+}
+
+/* Codes_SRS_UWS_CLIENT_01_213: [ A fragmented message consists of a single frame with the FIN bit clear and an opcode other than 0, followed by zero or more frames with the FIN bit clear and the opcode set to 0, and terminated by a single frame with the FIN bit set and an opcode of 0. ]*/
```
Comment (jebrando) [inline tests/uws_client_ut/uws_client_ut.c]: >Codes [](start = 3, length = 5)

Can you please change these to Tests instead of Codes.

Comment (blico) [reply tests/uws_client_ut/uws_client_ut.c]: done

File: src/uws_client.c (hunk 2 above)
Comment (blico) [reply src/uws_client.c]: added process_frame_fragment() function

**Comment by blico** [comment]
@jebrando I addressed your feedback in my latest commit. Let me know if you have any other feedback.

**Pull request merged**